Include original state in learner response report

xitij2000 opened the pull request:
A [previous enhancement](https://github.com/edx/edx-platform/pull/17776) to the learner response report removed the original state parameter in favour of a cleaner state. This work restores the original state while still including the original enhancements. 

Also in this PR:
- The data provided by the xblock report generator is now available in columns in the final CSV instead of as a JSON string in a single column. 
- For certain kinds of problems the correct answer is also included in the final report under the "Correct Answer" column.  

**JIRA tickets**: https://openedx.atlassian.net/browse/EDUCATOR-3143

**Sandbox URL**: https://pr18512.sandbox.opencraft.hosting/

**Testing instructions**:
Run included tests with: 
``pytest lms/djangoapps/instructor_task/tests/test_tasks_helper.py::TestProblemResponsesReport``

**Sample Output**:


username | block_key | title | Question | Answer | state | location | Answer ID
-- | -- | -- | -- | -- | -- | -- | --
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@c554538a57664fac80783b99d9d6da7c | Pointing on a Picture | Which animal is a kitten? | [529,186] | {"correct_map": {"c554538a57664fac80783b99d9d6da7c_2_1": {"hint": "", "hintmode": null, "correctness": "correct", "msg": "", "answervariable": null, "npoints": null, "queuestate": null}}, "input_state": {"c554538a57664fac80783b99d9d6da7c_2_1": {}}, "last_submission_time": "2018-07-03T18:40:53Z", "attempts": 1, "score": {"raw_earned": 1, "raw_possible": 1}, "done": true, "student_answers": {"c554538a57664fac80783b99d9d6da7c_2_1": "[529,186]"}, "seed": 1} | Homework - Question Styles > Pointing on a Picture > Pointing on a Picture | c554538a57664fac80783b99d9d6da7c_2_1
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@d2e35c1d294b4ba0b3b1048615605d2a | Drag and Drop | Question 1 | [{"2":[430,178.0833282470703]},{"3":[190,162.0833282470703]},{"5":[25,141.0833282470703]},{"6":[281,154.0833282470703]},{"9":[548,206.0833282470703]}] | {"score": {"raw_possible": 1, "raw_earned": 0}, "correct_map": {"d2e35c1d294b4ba0b3b1048615605d2a_2_1": {"hint": "", "hintmode": null, "correctness": "incorrect", "msg": "", "answervariable": null, "npoints": 0, "queuestate": null}}, "input_state": {"d2e35c1d294b4ba0b3b1048615605d2a_2_1": {}}, "last_submission_time": "2018-07-03T18:41:08Z", "attempts": 1, "seed": 1, "done": true, "student_answers": {"d2e35c1d294b4ba0b3b1048615605d2a_2_1": "[{\"2\":[430,178.0833282470703]},{\"3\":[190,162.0833282470703]},{\"5\":[25,141.0833282470703]},{\"6\":[281,154.0833282470703]},{\"9\":[548,206.0833282470703]}]"}} | Homework - Question Styles > Drag and Drop > Drag and Drop | d2e35c1d294b4ba0b3b1048615605d2a_2_1
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@a0effb954cca4759994f1ac9e9434bf4 | Multiple Choice Questions | Which piece of furniture is built for sitting? | a chair | {"correct_map": {"a0effb954cca4759994f1ac9e9434bf4_4_1": {"hint": "", "hintmode": null, "correctness": "correct", "msg": "", "answervariable": null, "npoints": null, "queuestate": null}, "a0effb954cca4759994f1ac9e9434bf4_2_1": {"hint": "", "hintmode": null, "correctness": "incorrect", "msg": "", "answervariable": null, "npoints": null, "queuestate": null}, "a0effb954cca4759994f1ac9e9434bf4_3_1": {"hint": "", "hintmode": null, "correctness": "correct", "msg": "", "answervariable": null, "npoints": null, "queuestate": null}}, "input_state": {"a0effb954cca4759994f1ac9e9434bf4_4_1": {}, "a0effb954cca4759994f1ac9e9434bf4_2_1": {}, "a0effb954cca4759994f1ac9e9434bf4_3_1": {}}, "last_submission_time": "2018-07-03T18:41:13Z", "attempts": 3, "score": {"raw_possible": 3, "raw_earned": 2}, "done": true, "student_answers": {"a0effb954cca4759994f1ac9e9434bf4_4_1": ["choice_0", "choice_2"], "a0effb954cca4759994f1ac9e9434bf4_2_1": "yellow", "a0effb954cca4759994f1ac9e9434bf4_3_1": "choice_2"}, "seed": 1} | Homework - Question Styles > Multiple Choice Questions > Multiple Choice Questions | a0effb954cca4759994f1ac9e9434bf4_3_1
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@Sample_Algebraic_Problem | Mathematical Expressions | Question 1 | A*x^2 + sqrt(y) | {"correct_map": {"Sample_Algebraic_Problem_2_1": {"hint": "", "hintmode": null, "correctness": "correct", "msg": "", "answervariable": null, "npoints": null, "queuestate": null}}, "input_state": {"Sample_Algebraic_Problem_2_1": {}}, "last_submission_time": "2018-07-03T18:42:00Z", "attempts": 1, "score": {"raw_earned": 1, "raw_possible": 1}, "done": true, "student_answers": {"Sample_Algebraic_Problem_2_1": "A*x^2 + sqrt(y)", "Sample_Algebraic_Problem_2_1_dynamath": "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\r\n  <mstyle displaystyle=\"true\">\r\n    <mi>A</mi>\r\n    <mo>&#x22C5;</mo>\r\n    <msup>\r\n      <mi>x</mi>\r\n      <mn>2</mn>\r\n    </msup>\r\n    <mo>+</mo>\r\n    <msqrt>\r\n      <mrow>\r\n        <mi>y</mi>\r\n      </mrow>\r\n    </msqrt>\r\n  </mstyle>\r\n</math>"}, "seed": 1} | Homework - Question Styles > Mathematical Expressions > Mathematical Expressions | Sample_Algebraic_Problem_2_1
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@Sample_ChemFormula_Problem | Chemical Equations |   |   | {"score": {"raw_earned": 0, "raw_possible": 1}, "seed": 1, "input_state": {"Sample_ChemFormula_Problem_2_1": {}}} | Homework - Question Styles > Chemical Equations > Chemical Equations |  
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@75f9562c77bc4858b61f907bb810d974 | Numerical Input |   |   | {"score": {"raw_earned": 0, "raw_possible": 3}, "seed": 1, "input_state": {"75f9562c77bc4858b61f907bb810d974_2_1": {}, "75f9562c77bc4858b61f907bb810d974_3_1": {}, "75f9562c77bc4858b61f907bb810d974_4_1": {}}} | Homework - Question Styles > Numerical Input > Numerical Input |  
edx | block-v1:edX+DemoX+Demo_Course+type@problem+block@0d759dee4f9d459c8956136dbde55f02 | Text Input |   |   | {"score": {"raw_earned": 0, "raw_possible": 1}, "seed": 1, "input_state": {"0d759dee4f9d459c8956136dbde55f02_2_1": {}}} | Homework - Question Styles > Text input > Text Input |  


**Reviewers**
- [ ] @bradenmacdonald 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
JOURNALS_ENABLED: false
# copied from https://github.com/edx/configuration/blob/cf4d221d384ed396a3008c31487f385544ef08e7/util/jenkins/ansible-provision.sh#L328-L338
JOURNALS_URL_ROOT: "https://journals-{{ EDXAPP_LMS_BASE }}"
JOURNALS_API_URL: "https://journals-{{ EDXAPP_LMS_BASE }}/api/v1/"
JOURNALS_DISCOVERY_SERVICE_URL: "https://discovery-{{ EDXAPP_LMS_BASE }}"
JOURNALS_LMS_URL_ROOT: "https://{{ EDXAPP_LMS_BASE }}"
JOURNALS_SOCIAL_AUTH_REDIRECT_IS_HTTPS: true
JOURNALS_DISCOVERY_API_URL: "{{ JOURNALS_DISCOVERY_SERVICE_URL }}/api/v1/"
JOURNALS_DISCOVERY_JOURNALS_API_URL: "{{ JOURNALS_DISCOVERY_SERVICE_URL }}/journal/api/v1/"
JOURNALS_ECOMMERCE_BASE_URL: "{{ ECOMMERCE_ECOMMERCE_URL_ROOT }}"
JOURNALS_ECOMMERCE_API_URL: "{{ JOURNALS_ECOMMERCE_BASE_URL }}/api/v2/"
JOURNALS_ECOMMERCE_JOURNALS_API_URL: "{{ JOURNALS_ECOMMERCE_BASE_URL }}/journal/api/v1"
journals_create_demo_data: false

xqueue_source_repo: https://github.com/edx/xqueue
xqueue_version: master
CERTS_REPO: https://github.com/edx/edx-certificates
certs_version: master
```

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-2503](https://openedx.atlassian.net/browse/OSPR-2503) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

mduboseedx commented:
@edx/educator-product This is making a small change to the one of the Data Download Reports and includes the original "state" parameter. Could you give this a quick review?

xitij2000 commented:
@mduboseedx I don't have access to the actual ticket but I think this addresses [EDUCATOR-3143](https://openedx.atlassian.net/browse/EDUCATOR-3143) among other things.

bradenmacdonald commented on lms/djangoapps/instructor_task/tasks_helper/grades.py:
```diff
@@ -639,12 +643,15 @@ def _build_student_data(cls, user_id, course_key, usage_key_str):
                         break
 
-        return student_data
+        return student_data, list(student_data_keys)
```
The docstring says this returns a `Set` but here it's returning a `list`

xitij2000 replied:
Updated the docstring.

xitij2000 commented:
@bradenmacdonald I've made the order somewhat stable: 
username, title, location, ...generated columns in sorted order..., block_key, state
So the less reader-friendly columns are at the end, and the more relevant ones are at the beginning. 

This was somewhat simple to do,  but unless we have a stable set of columns returned by ``generate_report_data`` it's hard to have a very stable order. Other blocks might return results more approprate for their data.

bradenmacdonald commented on lms/djangoapps/instructor_task/tasks_helper/grades.py:
```diff
@@ -651,7 +651,16 @@ def _build_student_data(cls, user_id, course_key, usage_key_str):
+            ['username', 'title', 'location'] +
+            sorted(student_data_keys) +
+            ['block_key', 'state']
+        )
```
Very nice solution, thanks.

xitij2000 commented:
jenkins run bokchoy

awaisdar001 commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -185,7 +185,7 @@ def async_migrate_transcript(self, course_key, **kwargs):   # pylint: disable=un
             sub_tasks = []
             video_location = unicode(video.location)
-            for lang in all_transcripts.keys():
+            for lang in all_transcripts:
```
why this changed from `.keys()`?

xitij2000 replied:
It was showing up as a new warning for some reason: 
https://build.testeng.edx.org/job/edx-platform-quality-flow-pr/5858/warnings52Result/new/

awaisdar001 replied:
Ok, got it.

awaisdar001 commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -495,6 +495,18 @@ def get_answer_ids(self):
         return answer_ids
 
+    def find_correct_answer_text(self, answer_id):
+        xml_elems = self.tree.xpath('//*[@id="' + answer_id + '"]')
```
Can we add a doc string for this method?

awaisdar001 replied:
Can we add more readable variable name? 
`xml_elems` => `xml_elements`?

awaisdar001 commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -495,6 +495,18 @@ def get_answer_ids(self):
+        xml_elems = self.tree.xpath('//*[@id="' + answer_id + '"]')
+        if not xml_elems:
+            return
+        xml_elem = xml_elems[0]
```
`xml_elem`=>`xml_element`

awaisdar001 commented on common/lib/xmodule/xmodule/capa_module.py:
```diff
@@ -393,13 +393,17 @@ def generate_report_data(self, user_state_iterator, limit_responses=None):
-                })
+                }
+                if correct_answer_text is not None:
+                    report[_("Correct Answer")] = correct_answer_text
```
👍

awaisdar001 commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -496,16 +496,26 @@ def get_answer_ids(self):
+
+        Returns:
+            str: A string containing the correct answer or a comma-separated
+                 list of the correct answers.
```
I believe it is `comma-separated string` than a `list`?

xitij2000 replied:
Yup, sorry that was a bit ambiguous.

awaisdar001 commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -659,6 +659,37 @@ def test_find_answer_text_choices(self, answer_id, choice_id, answer_text):
+    )
+    @ddt.unpack
+    def test_find_correct_answer_text_choices(self, answer_id, answer_text):
+        problem = new_loncapa_problem(
```
Please, can you add what scenario is being tested here?

xitij2000 commented:
@awaisdar001 @Rabia23  I've updated the code to improve the variable names and added the docstring.

awaisdar001 commented on lms/djangoapps/instructor_task/tasks_helper/grades.py:
```diff
@@ -661,14 +677,17 @@ def generate(cls, _xmodule_instance_args, _entry_id, course_id, task_input, acti
-        features = ['username', 'title', 'location', 'block_key', 'state']
-        header, rows = format_dictlist(student_data, features)
+        for data in student_data:
+            for key in student_data_keys:
```
what about the oneliner? as it is only used for setting default values.
`[data.setdefault(key, '') for key in student_data_keys]`

I am open to this change. its up to you.

xitij2000 replied:
I think that would be a bit less clear, would prefer to keep it the way it is. No strong feeling either way though!

xitij2000 commented:
jenkins run bokchoy

bradenmacdonald commented:
@awaisdar001 Yep I could do that.

@xitij2000 Looks like there's a quality issue after the rebase though :/

xitij2000 commented:
@bradenmacdonald  On it.

xitij2000 commented:
@bradenmacdonald  Apparently there is an arbitrary limit of 973 stylelint errors that are allowed, and for some reason, it's complaining about that now. 

So I've fixed applied automated fixes for all style files and manually cleared out the error from the file that had the errors. It's now down to 718 errors but makes the PR too large. 

How should I tackle this? Should I make a separate PR for all the fixes? Should I just include one fixed file in this PR just to get it over the threshold? For now, I've included all of them. 

CC: @awaisdar001

bradenmacdonald commented:
@xitij2000 That sort of unrelated quality count issue can happen when someone else merges a PR without rebasing first. It's just the luck of the draw that it's now affecting you. I would suggest including just one fix in this PR to get it over the threshold, and submit your other fixes in a separate PR.

Apparently one of the CSS whitespace changes actually causes a python test to fail, so omit that one from both.

xitij2000 commented:
@bradenmacdonald  I removed all but one file that I fixed. 

I've created [a new PR](https://github.com/edx/edx-platform/pull/18545) with the rest of the changes, that can be merged if useful.

xitij2000 commented:
jenkins run python

xitij2000 commented:
jenkins run bokchoy

xitij2000 commented:
jenkins run bokchoy

xitij2000 commented:
@awaisdar001 I think there might be issues with the ``acceptance.tests.studio.test_studio_help.LibraryHelpTest.test_library_content_side_bar_help`` test. 

I don't see how it's related to the PR, but it's starting to fail more often than not, due to timeouts. 

Evidence:

- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56109/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56087/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56218/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56214/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56213/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56154/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56151/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/

xitij2000 commented:
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56109/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56087/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56218/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56214/testReport/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56213/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56154/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/56151/testReport/junit/acceptance.tests.studio.test_studio_help/LibraryHelpTest/test_library_content_side_bar_help/

edx-webhook commented:
@xitij2000 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18512&entry.1484655318=https://github.com/xitij2000&entry.752974735=2018-07-03+19:32&entry.1917517419=2018-07-09+23:14&entry.2133058324=Yes) so we can improve your experience in the future.

nedbat commented:
I've cherry-picked the first two commits (the substantive ones) onto Hawthorn.

bradenmacdonald commented:
Thanks @nedbat !

xitij2000 commented:
Thanks!
